pkg/api/service.go

```diff
@@ -55,15 +64,63 @@ func (s *Service) Close() {
 
 func (s *Service) BatchSubscribeEnvelopes(
 	req *message_api.BatchSubscribeEnvelopesRequest,
-	server message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
+	stream message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
 ) error {
-	return status.Errorf(codes.Unimplemented, "method BatchSubscribeEnvelopes not implemented")
+	log := s.log.Named("subscribe")
+
+	// Send a header (any header) to fix an issue with Tonic based GRPC clients.
+	// See: https://github.com/xmtp/libxmtp/pull/58
+	err := stream.SendHeader(metadata.Pairs("subscribed", "true"))
+	if err != nil {
+		return status.Errorf(codes.Internal, "could not send header: %v", err)
+	}
+
+	requests := req.GetRequests()
+	if len(requests) == 0 {
+		return status.Errorf(codes.InvalidArgument, "missing requests")
+	}
+
+	ch, err := s.subscribeWorker.addListeners(requests)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "invalid subscription request: %v", err)
+	}
+
+	var streamLock sync.Mutex
+	for exit := false; !exit; {
+		select {
+		case envs, open := <-ch:
+			if open {
+				func() {
+					streamLock.Lock()
+					defer streamLock.Unlock()
+					err := stream.Send(&message_api.BatchSubscribeEnvelopesResponse{
+						Envelopes: envs,
+					})
+					if err != nil {
+						log.Error("sending envelope to subscribe", zap.Error(err))
+					}
+				}()
+			} else {
+				// TODO(rich) Recover from backpressure
+				log.Info("stream closed due to backpressure")
+				exit = true
+			}
+		case <-stream.Context().Done():
+			log.Debug("stream closed")
+			exit = true
+		case <-s.ctx.Done():
+			log.Info("service closed")
+			exit = true
+		}
+	}
+	return nil
 }
 
 func (s *Service) QueryEnvelopes(
 	ctx context.Context,
 	req *message_api.QueryEnvelopesRequest,
 ) (*message_api.QueryEnvelopesResponse, error) {
+	log := s.log.Named("query")
```

Comment:
Instead of taking over the name of the logger, it might be more readable to have the struct hold a logger named `api` and to add context for each method using `With` for each endpoint. Then we can group together all the API handlers in DataDog.